DM-45371: technote describing possible cell-based coadd file layouts

TallJimbo opened the pull request.

TallJimbo committed:
Re-whitespace abstract and fix extraneous word.

TallJimbo committed:
Add intros, outline, and metadata sections.

TallJimbo committed:
Add sections for binary tables and per-cell HDUs.

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+As a result, I expect us to need to write some low-level code of our own to do subimage reads over object stores, though this is not specific to cell-based coadds: we need this for all of our image data products.
+Whether to contribute more general code upstream (probably to Astropy) or focus only on reading the specific files we write ourselves is an important open question; the general solution would be very useful the community, but its scope is unquestionably larger.
+
+Finally, the WCS of any single cell (or the inner-cell stitched image) will be simple and can be exactly represented in FITS headers (unlike the WCSs of our single-epoch images).
```
To be very clear, the WCS of a single cell is simply that of the tract. The layout has only one WCS and each cell is simply some subimage within that WCS. This feature makes the overlap pixels be literally the same geometrically, means that the images can be stitched together correctly, etc.

gpdf replied:
That's _conceptually_ true but if the cells are stored independently of each other in the FITS file, they do technically have separate WCSes and will at least need different CRPIXn values in order to participate in the tract-global WCS correctly.

(Of course the overlap pixels don't have the same _values_, since there's a different input image selection for each cell.)

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+Whether to contribute more general code upstream (probably to Astropy) or focus only on reading the specific files we write ourselves is an important open question; the general solution would be very useful the community, but its scope is unquestionably larger.
+
+Finally, the WCS of any single cell (or the inner-cell stitched image) will be simple and can be exactly represented in FITS headers (unlike the WCSs of our single-epoch images).
+An additional FITS WCS could be used to describe the position of a single-cell image within a larger grid.
```
This position is simply a pixel location and the dimension of the image.

gpdf replied:
I think Jim is referring to the Rubin convention of having an "alternative WCS" (we usually use WCS `A`) that provides the mapping from local FITS 1-based pixel coordinates to tract-level 0-based coordinates.

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+
+In fact, even without compression, the binary table layout would require writing our code just to solve the problem of subimage reads over object stores, since Astropy cannot do efficient table subset reads and CFITSIO cannot do object store reads.
+
+Per-HDU Cells
```
There will be a lot of cells and I've seen serious performance issues in FITS files with many many HDUs. I'd be wary of this one.

TallJimbo replied:
My mental model is that this is because you have to seek through the whole file to find the byte addresses where each HDU starts, and I think an address table will fix this _for readers that use it_.  But it's still a problem for other readers that are unaware of it.

beckermr replied:
I am not sure if I ever understood properly what the issues was. @esheldon, did we get to the bottom of this for fitsio?

esheldon replied:
That might have been a python `fitsio` issue only, not an issue with `cfitsio`.   I think the problem was fixed a while back, but if you think this persists please make an issue.

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+
+This section will cover potential ways to lay out the image data - the 5+ planes that share a common grid, as well as the PSF images - across multiple FITS HDUs.
+
+Binary Table With Per-Cell Rows
```
I think you are aware, but the FITS tile compression code fpack ends up with a binary table of compressed images. In DES, we get pretty good (lossy) compression with the tiling working per cell. We found pretty apparent issues with lossy compression when the tile compression boundaries crossed cell boundaries because the noise levels in the images change.

timj replied:
It really makes me wonder why we aren't using HDF5 and making use of a hierarchy and native compression rather than trying to shoe horn this into FITS.

TallJimbo replied:
The way I see it, we can _also_ do HDF5, but we really can't _just_ do HDF5, according to our requirements.  Maybe there's some scheme where we only store the HDF5 and convert to FITS when streaming bytes in various places, but I'm not at all convinced that removes enough constraints from the FITS version to make it easier to implement overall.

beckermr replied:
> It really makes me wonder why we aren't using HDF5 and making use of a hierarchy and native compression rather than trying to shoe horn this into FITS.

This is an interesting perspective. Given that FITS already provides a solution that generates this kind of data structure internally and has all of the support to read subsets etc., it seems like it'd be shoe horning to put this into HDF5. One would have to build all of the logic to interact with the table yourself when something already exists and is a defacto community standard.

It seems like the only issue with simply using what FITS has is that one cannot use object stores / non-posix I/O to read subsets of the file. That seems like the kind of use case that is specific to DM, is solvable with custom code that can be written just for DM's use, and results in reusing essentially everything FITS already offers.

timj replied:
Forcing cell-based coadds into FITS when the thing you end up with is essentially unusable for ds9/firefly/astropy seems like we are stretching the usefulness of the requirement we have to use FITS. If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough. cc/ @gpdf

beckermr replied:
**To be very clear, the cells with their buffers are the usable images.** Any algorithm one runs on them will typically be run on the full cells with buffers and then info from the buffers is discarded at the end.

TallJimbo replied:
I think there are two cases:

- if you want to do something convolution-like, you want the full cells;
- if you want to do forward modeling (or naive things that don't care about the PSF), you don't care about the outer regions.  Instead you just render a piecewise model with the piecewise PSF and compare that to the piecewise data.

While both are important, I actually think the latter will be much more common.

arunkannawadi replied:
Matt and Tim mean different things by the term 'usable'. Matt is referring to specific algorithms being able to run as usable, whereas Tim is referring to third-party tools able to interpret, stitch and visualize as 'usable'

> If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough.

Interesting. Do we have a service to do a bulk data transfer? Such a conversion _could_ happen in the process, since we don't want to store the trimmed/stitched image at USDF and the destination may not have Science Pipelines installed to read our custom datastructure.

beckermr replied:
The thing with item 2 above is that as soon as you think about how to parallelize a forward modeling algorithm, you'll end up back in the case of using some sort of buffer. The overlaps are meant to enable these things and not an annoyance that we have to work around.

TallJimbo replied:
Cells are one parallelization axis, sure.  But they're not the only one; parellelizing over objects is another, and we'll have enough data that we can keep all the machines plenty hot parallelizing only over patches.  I'd generally say that parallelizing over cells is best only if you have your own reason to do per-cell detection and hence have no choice but to deal with the problem of resolving duplicate detections later.  Otherwise it's a lot cleaner to combine data at the pixel level, especially if you're starting from our main Object table's detections.

TallJimbo replied:
> Do we have a service to do a bulk data transfer?

We don't, and "bulk" has implications that we don't want to get into.  But services that transform or augment images as they are downloaded is something we've talked about for a long time.

My main concern with only providing FITS via a service is that I think there's still a lot of work defining the format of the FITS file we return.  Stitching together the inner cells is not enough to make a coadd look just like a single-epoch image, at least not when you're thinking about the on-disk data model rather than a Python API.  I don't want to put a lot of effort into an HDF5 format for cell-based coadds (especially not if I'm also putting a lot of effort into FITS+JSON serialization of other image data products), only to have the scope of what people want back from the FITS-transformer service creep up to near parity with what's in the HDF5.

gpdf replied:
@timj wrote:
> Forcing cell-based coadds into FITS when the thing you end up with is essentially unusable for ds9/firefly/astropy seems like we are stretching the usefulness of the requirement we have to use FITS. If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough.

I think that the binary table model is a reasonable one.  It solves one of the biggest problems for interoperable, non-mission-specific interpretation of FITS files with enormous numbers of HDUs: there's nothing in the FITS standard, and precious little in the "registered conventions", for having a "table of contents" extension that tells you what's in the file.

By comparison, putting the subimages in a table solves this problem for free.  You can put all sorts of metadata into additional columns in the table, and a reasonable application will give a user the ability to browse by this metadata.

It's true that _today_ if you handed a big Green Bank style table to Firefly or Aladin you might not get acceptable UX, but that's because there's been little pressure on tools in this area to catch up with what's already in the standard.  

It's important to recognize that Rubin isn't the only project with this need, and cell-based coadds aren't even the only application for it within Rubin: we also have an equally important requirement to be able to create multi-epoch bulk cutout files, both for cutouts around many objects, and for cutouts around one (possibly moving) object in many epochs.

So we are among friends, as it were, in looking for standards-conformant improvements in tool behavior.

gpdf replied:
> > If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough.
> 
> Interesting. Do we have a service to do a bulk data transfer? Such a conversion could happen in the process, since we don't want to store the trimmed/stitched image at USDF and the destination may not have Science Pipelines installed to read our custom datastructure.

We have a robust framework for retrieving "related data" in the RSP.  When someone queries for coadded images, they'll get a list of images, most naturally, in the way they are chunked in the Butler (e.g., currently, one row per patch in the query result).

The way the image itself is then retrieved actually allows for N different products to be returned associated with that row in the query result.  So you can return the full cell-based coadd FITS file verbatim from the backing store, or you can return a stitched-together simple FITS file for the whole area it covers, or you can return a PNG preview of it, or a table of its provenance, or pretty much anything else we can think of.

These additional related products can be things that already exist statically, or things that are created on the fly by services from the data on the backing store.

gpdf replied:
... and modern IVOA-flavored tools like Firefly and Aladin already know how to follow these links and present these options to users.

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+Exploded Images
+---------------
+
+TODO
```
The MEDS format is basically exploded images but with them all unraveled into a single 1D set of pixels. We then set the lossy compression tile boundaries to correspond to the exploded cells with one per tile.

beckermr replied:
To be clear, we store one exploded image per data type being stored (e.g., one image HDU, one mask HDU etc.) and then they all share common metadata stored in a binary table HDU in the file.

gpdf replied:
I'm not familiar with this format, @beckermr .  Do you have a reference?

arunkannawadi replied:
https://github.com/esheldon/meds/wiki/MEDS-Format

I do not know if this is up to date with what has been done in practice though.

beckermr commented on index.rst:
```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+While some global information will go into FITS headers (certainly the WCS and some identifiers), we do not want to assume that all global metadata can be neatly represented in the limited confines of a FITS header.
+A single-row binary table is another option, but we will likely instead adopt the approach recently proposed for other Rubin image data products on RFC-1030: embedding a JSON document as a byte array in a FITS extension HDU.
+
+A binary table with per-cell rows is a natural fit for the fixed-schema per-cell information, especially if the image data layout already involves a binary table with per-cell rows.
+But if we're embedding a JSON document in the FITS file anyway, it might make more sense to store this information in JSON as well; this will let us share code, documentation, and serialization for more complex objects with other Rubin image data products, and that includes sharing the machinery for managing schema changes schema documentation.
```
Have you seen performance issues with JSON in FITS? 

We do this very rarely in DES and it is a potentially lossy operation, especially with respect to floating point numbers etc.

TallJimbo replied:
You do have to be very careful with the floating point numbers (NaNs and Infs especially), but the libraries we're using (mostly Pydantic) seem to be pretty good about that.

The JSON-in-FITS stuff is extremely new and we don't have any benchmarks of that specifically.  But I don't think the combination raises any red flags, and we've been pretty happy with JSON serialization even in some very big data structures (not YAML, though, avoid YAML like the plague if you care about performance!).

esheldon replied:
My 2 cents:  As an end user I would prefer binary table if the metadata has a "flat" (non hierarchical) structure and has FITS compatible data types.

Using binary blobs makes data exploration/browsing more difficult for anyone who is not using DM specific "data loaders" etc.

TallJimbo committed:
Add sections on data cubes and exploded images.

TallJimbo committed:
Add section on stitched images and hybrid approaches.

TallJimbo committed:
Proofreading edits and PR comment clarifications.

TallJimbo committed:
Delete template text.

arunkannawadi commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+A single-row binary table is another option, but we will likely instead adopt the approach recently proposed for other Rubin image data products on RFC-1030: embedding a JSON document as a byte array in a FITS extension HDU.
+
+A binary table with per-cell rows is a natural fit for the fixed-schema per-cell information, especially if the image data layout already involves a binary table with per-cell rows.
+But if we're embedding a JSON document in the FITS file anyway, it might make more sense to store this information in JSON as well; this will let us share code, documentation, and serialization for more complex objects with other Rubin image data products, and that includes sharing the machinery for managing schema changes schema documentation.
```
```suggestion
But if we're embedding a JSON document in the FITS file anyway, it might make more sense to store this information in JSON as well; this will let us share code, documentation, and serialization for more complex objects with other Rubin image data products, and that includes sharing the machinery for managing schema changes and schema documentation.
```

arunkannawadi replied:
Missing `and` ?

arunkannawadi commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
 
-   Rubin's deep coadds will be built in on a grid of small cells, in which each cell has an approximately constant PSF.  Cells will have "inner regions" that can be stitched together to form the full coadd, but they will also have outer regions that overlap (neighboring cells will have their own versions of some of the same pixels), in order to allow convolutions and other operations that require ...
+   Rubin's deep coadds will be built on a grid of small cells, in which each cell has an approximately constant PSF.
+   Cells will have "inner regions" that can be stitched together to form the full coadd, but they will also have outer regions that overlap (neighboring cells will have their own versions of some of the same pixels), in order to allow convolutions and other operations that require padding to be performed rigorously cell by cell.
```
```suggestion
   Cells will have "inner regions" that can be stitched together to form the full (patch-sized) coadd, but they will also have outer regions that overlap (neighboring cells will have their own versions of some of the same pixels), in order to allow convolutions and other operations that require padding to be performed rigorously cell by cell.
```

arunkannawadi commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+This avoids the problem with per-HDU overheads, and it makes an address table much less important, as there are many fewer HDUs.
+It also allows compression tiles that comprise multiple cells.
+
+It does not allow us to represent the on-sky locations of cells using FITS WCS, however, and this is probably enough to rule it out.
```
Are you referring to the integer offsets that indicate the position of the cell?

TallJimbo replied:
Yes, more or less.  We could have a WCS that describes one cell, but I don't think there's any way to write a FITS WCS that puts different slices of a data cube in different places.

timj replied:
I don't think FITS-WCS can do it. I'm pretty sure AST can with switchMaps.

timj replied:
No idea if gwcs can.

gpdf replied:
I believe there might be a way to do this with table-lookup WCSes, but they are not uniformly well supported by readers.  I would have to read the FITS standard very closely to determine whether this would also require us to do all the spherical projection ourselves, which would require us to be careful about precision issues in the (piecewise linear) table lookups, especially near the coordinate poles.

I am not endorsing this idea!  But if this storage scheme were the winner for other reasons, it's something we could research.

TallJimbo replied:
This storage scheme definitely isn't winning for other reasons, so I don't think it's worth anyone's time to look into it further.

arunkannawadi commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+Assuming we don't care about FITS WCS support for the overlap regions, the main problem with this approach is complexity in two places:
+
+- When writing, we'd need to quantize the outer cell image first, and then slice the image into its inner-cell and overlap-region sections, and only then compress the quantized values.
```
Can you clarify (either as a comment here, or in the document) if you mean to compress not just the inner cell, but _also_ the overlap regions? And if so, can just the inner cell be decompressed and read without ever having to access the overlap regions?

TallJimbo replied:
I intend to quantize the full cell, then compress both the inner regions and overlap regions, but to compress them separately (and possibly compress multiple cells' inner regions together, and especially multiple cells' overlap regions together).  So it would be possible to decompress just the inner cell region.

beckermr replied:
So to be very clear, you need to quantize separately per cell. Compression together is fine though. This is what is in your text above but I wanted to state it very plainly to ensure we're on the same page.

TallJimbo committed:
PR review edits from Arun

Co-authored-by: Arun Kannawadi  <kj.arun.kj@gmail.com>

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+- Astropy can do image tile compression with efficient subimage reads on POSIX filesystems, and it can do uncompressed reads (including efficient subimage reads) against both POSIX and object stores via `fsspec`, but it delegates to vendored CFITSIO code for its tile compression code and does not support the combination of these two.
+  Astropy cannot do subset reads of FITS binary tables, even on POSIX filesystems.
+
+- I am not aware of any library that can do FITS binary table compression at all.
```
I think in fact CFITSIO (C) and nom.tam.fits (Java) both support it.  It appears that the CFITSIO documentation is behind the code.

Note that CFITSIO was only recently formally released as open-source code on GitHub (there were semi-bootleg versions of it previously) and people may not be entirely familiar with the latest version.

Some links:

https://github.com/HEASARC/cfitsio/blob/56640a39c59148b5143d8dd0833cce6a88a55d8c/fitsio.h#L2069-L2070

https://github.com/HEASARC/cfitsio/blob/56640a39c59148b5143d8dd0833cce6a88a55d8c/imcompress.c#L8704

gpdf replied:
Currently *lossy* table compression is not supported.

> The permitted algorithms for compressing BINTABLE columns are ’RICE 1’, ’GZIP 1’, and ’GZIP 2’ (plus ’NOCOMPRESS’), which are lossless and are described in Sect. 10.4. Lossy compression could be allowed in the future once a process is defined to preserve the details of the compression.

TallJimbo replied:
Fixed.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+Binary Table With Per-Cell Rows
+-------------------------------
+
+Since FITS binary tables can have array columns with arbitrary shapes, the most intuitive layout for cell-based image data is probably a single binary table HDU that has each image plane as a separate 2-d array column.
```
Agreed.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+Since FITS binary tables can have array columns with arbitrary shapes, the most intuitive layout for cell-based image data is probably a single binary table HDU that has each image plane as a separate 2-d array column.
+A FITS WCS can be stored with each row, using "Green Bank" convention to translate header keys to table columns; this may be sufficient for third-party readers to fully interpret our files and even stitch together cells on the fly (in the case of image viewers).
+
+It's not clear whether any extant FITS viewers could *already* do this in an intuitive way, but by relying on an existing convention it'd be reasonable for them to add it, even if they were otherwise disinclined to add support for observatory-specific file formats.
```
I doubt that any existing viewer _already_ has the stitching capability in this use case, but as Jim says it's a realistic expectation that they might be developed to do that, particularly if this becomes a more widely used organization for data in the community.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+The main problem with binary table layouts is compression support.
+The latest FITS standard does include (in section 10.3) a discussion of "tiled table compression", which seems like it'd be sufficient, as long as compressing one cell at a time is enough to get a good compression ratio (this is unclear).
+Unlike image tile compression, binary table tile compression doesn't support lossy compression algorithms or dithered quantization, but it would still be possible to do our own non-dithered quantization and use ``BZERO`` and ``BSCALE`` to record the transformation from integer back to floating-point.
```
Agreed.  This should work.

The FITS standard also explicitly says that the door is open to lossy compression algorithms for tables if a specific technical problem is solved.  So we can try to pursue that further as we move along this road.

TallJimbo replied:
Given that `BZERO` and `BSCALE` don't provide a way to subtractive dithering (and after from hearing from @beckermr that this was important in practice), I now don't believe this approach would meet our needs.  I've left the discussion of the possibility in, but have reframed it as an incomplete solution.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+The main problem with binary table layouts is compression support.
+The latest FITS standard does include (in section 10.3) a discussion of "tiled table compression", which seems like it'd be sufficient, as long as compressing one cell at a time is enough to get a good compression ratio (this is unclear).
+Unlike image tile compression, binary table tile compression doesn't support lossy compression algorithms or dithered quantization, but it would still be possible to do our own non-dithered quantization and use ``BZERO`` and ``BSCALE`` to record the transformation from integer back to floating-point.
+The bigger problem is that there does not appear to be any implementations of it: there is no mention of it in either the CFITSIO or Astropy documentation (and even if an implementation does exist in, say, the Java ecosystem, we wouldn't be in a position to use it).
```
As mentioned above, it does appear that the current version of CFITSIO does support this.

On the Java side, which is important to Firefly, `nom.tam.fits` also claims to support it.

TallJimbo replied:
Fixed.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+While we've already discussed the fact that we'll probably need to implement some low-level FITS image tile compression code in order to do decompressed subimage reads with object stores anymore, the binary table compression situation is much more problematic:
+
+- tables are much more complicated than images;
+- we would have to implement writes ourselves, not just reads;
```
If, as stated elsewhere in the document, we can write with CFITSIO to Posix files, then all we have to implement at first is _reading_ via Astropy.

TallJimbo replied:
This has become a moot point because I was wrong about tiled table compression support in third-party libraries, and rewrote this section as a result.  But at the same time I've realized that we do need some amount of low-level write support work in order to do our own quantization, since Astropy does not support that, and the stuff in `afw` that does it is a mess (dependencies on undocumented and not-obviously-public CFITSIO calls) and not easy to call outside `afw` C++ code.  I've added some words on that in the latest revision.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+- tables are much more complicated than images;
+- we would have to implement writes ourselves, not just reads;
+- we would not have a reference implementation we could use for testing;
```
It looks like we have two; having a Java version _does_ help with this.

TallJimbo replied:
Fixed.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+- tables are much more complicated than images;
+- we would have to implement writes ourselves, not just reads;
+- we would not have a reference implementation we could use for testing;
+- if the standard has not seen real use, we stand a good chance of discovering uncovered edge cases or other defects;
```
That I completely agree with even though there are existing implementations.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+Another simple file layout is to put each image plane for each cell in a completely separate FITS image HDU.
+This is entirely compatible with FITS tile compression (though we'd almost certainly compress the entire HDU as one tile) and our goals for using FITS WCS.
+Stitching images from different HDUs into a coherent whole is probably a bit more likely for a third-party FITS viewer to support than images from different binary tables, but a flat list of HDUs for all cells and image planes provides a lot less organizational structure than a binary table (especially a single binary table) for third-party tools to interpret.
```
Yes, stitching across HDUs is supported by some FITS viewers that obey the not-very-well-standardized NOAO mosaic-focal-plane header conventions.  We would want to include those headers.  I'm not sure how clear it is how those headers would handle the overlap regions the way we want, but that's an answerable question.

I believe that Firefly does _not_ currently do this, but if we were to choose this representation, obviously we'd put that on the to-do list.

TallJimbo commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+The main problem with binary table layouts is compression support.
+The latest FITS standard does include (in section 10.3) a discussion of "tiled table compression", which seems like it'd be sufficient, as long as compressing one cell at a time is enough to get a good compression ratio (this is unclear).
+Unlike image tile compression, binary table tile compression doesn't support lossy compression algorithms or dithered quantization, but it would still be possible to do our own non-dithered quantization and use ``BZERO`` and ``BSCALE`` to record the transformation from integer back to floating-point.
+The bigger problem is that there does not appear to be any implementations of it: there is no mention of it in either the CFITSIO or Astropy documentation (and even if an implementation does exist in, say, the Java ecosystem, we wouldn't be in a position to use it).
```
It appears I was wrong about this: as @gpdf discovered, both CFITSIO and at least one major Java FITS implementation ([nom.tam.fits](https://github.com/nom-tam-fits/nom-tam-fits)) do support table compression (CFITSIO just hasn't updated their docs to reflect this).

But having looked more closely at the table compression spec, I'm less sanguine about introducing our own quantization convention for lossy compression here, for two reasons:

- Adding `BZERO` and `BSCALE` columns as per the Green Bank convention only works if there is only one image column in the binary table, i.e. we put each logical plane in its own binary table HDU (this seems generally true of the Green Bank convention: there's no provision for multiple image columns).  This is fine, but I think it removes one of the advantages of the binary table form.

- There'd be no way to record the "subtractive dithering" quantization (in which one adds a deterministic uniform random number before rounding to int, and subtracts it when decompressing).  And while we could invent one (probably a `ZDITHER0` column, by analogy with the tiled image compression header key), users would have to use our code to correctly decompress in that case.  While other file layouts would require users who don't use our code to do some work, none of them would require them to implement anything as low-level as decompression.  @beckermr, do you know if you used subtractive dithering in your lossy-compression configuration?  It'd be the `ZQUANTIZ` header key.

In fact, if we decide to do binary tables with one plane per HDU to work around the first problem, we could just make them the very binary tables used to implement tiled image compression, but add extra columns (which are explicitly permitted) for e.g. a Green Bank convention WCS or other metadata.  FITS libraries often provide the option to read a compressed image HDU as an image or as a binary table, and with this format we'd provide all information in a fairly FITS-native way in at least one of those two interpretations.  This does mean that the tiled image compression representation would correspond to the exploded image, of course.

beckermr replied:
We use SUBTRACTIVE_DITHER_2 which preserves zeros exactly. You need dithering to remove biases from the lossy compression and having zeros go through unchanged has lots of nice advantages.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+To account for chromatic PSF effects we may also store images that correspond to the derivatives of the PSF with respect to some proxy for object SED, in at least some bands.
+
+In addition, we'll need to store considerable structured metadata, including a WCS, information about the grid structure, coadded aperture corrections and wavelength-dependent throughput, the visit and detector IDs of the epochs that contributed to each cell, and additional information about the pixel uncertainty (some measure of typical covariance, and an approximately constant variance that eit ...
+
```
Add:

"It is highly desirable that non-Rubin-specific third-party image viewers be able to understand the WCS, display coordinates for pixels indicated by users, and accurately overlay marks on the image at the locations of detected objects or other targets."

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+Stitched Images
+---------------
+
+We expect most accesses to our coadd files to be uninterested in the redundant overlap pixel values - instead, most science users will be interested in stitching together the inner cells to form a mostly-seamless patch-level image.
```
I would say "many", perhaps, not "most", given the discussions of the value of the images for algorithmic processing vs. human inspection.

TallJimbo replied:
I've made that change and added my point on forward-modeling vs. data-convolution analysis needs from the other thread.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+
+Finally, the WCS of any single cell (or the inner-cell stitched image; they differ only in an integer offset) will be simple and can be exactly represented in FITS headers (unlike the WCSs of our single-epoch images).
+An additional FITS WCS could be used to describe the position of a single-cell image within a larger grid (this is just an integer offset).
+It is highly desirable that any images we store in our FITS files have both to allow third-party FITS readers to fully interpret them.
```
Add:
"In some cases we might be using corners of the current FITS standard ([v4.00](https://fits.gsfc.nasa.gov/standard40/fits_standard40aa-le.pdf)) that existing community tools do not yet fully implement.  We consider this preferable to using entirely Rubin-unique formats and conventions, and we are prepared to support the community to some extent in filling out these corners (e.g., by adding support to Astropy and Firefly).

We are aware of other projects' interest in FITS representations for large numbers of small-ish images with associated metadata and will keep them informed of what we are planning."

TallJimbo replied:
I'm a little uncomfortable with the parenthetical, as I don't think we'll actually have the effort to add support to at least Astropy on the early-science timescale.

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+We can in theory use a different approach for each image plane, though for the most part the arguments are the same for all image planes.
+The PSF images are the big exception: they are already intrinsically in their own coordinate system that doesn't have a meaningful WCS, and they are unlikely to ever be lossy compressed (I actually don't have any intuition for whether they'd have good lossless compression ratios).
+
+This makes data cube or exploded storage of PSFs quite attractive (binary tables too, at least if if we determine that we don't need to compress them at all), even if other image planes are stored in other ways.
```
Data cube storage of anything that isn't _naturally_ a cube makes it quite difficult to provide metadata about the individual planes beyond a single number (e.g., wavelength or time).

gpdf commented on index.rst:
```diff
@@ -4,9 +4,184 @@ File Formats and Layouts for Cell-based Coadds
+While some global information will go into FITS headers (certainly the WCS and some identifiers), we do not want to assume that all global metadata can be neatly represented in the limited confines of a FITS header.
+A single-row binary table is another option, but we will likely instead adopt the approach recently proposed for other Rubin image data products on RFC-1030: embedding a JSON document as a byte array in a FITS extension HDU.
+
+A binary table with per-cell rows is a natural fit for the fixed-schema per-cell information, especially if the image data layout already involves a binary table with per-cell rows.
```
I would encourage storing the fixed-schema data as a table even if it is also in the JSON.

gpdf reviewed (commented):
It's a great summary of the issues.  My main contribution is to note that table-compression implementations _are_ available.

I've made a number of other comments.  I'll write separately on the RFC about the proposed decision itself, which isn't addressed in this PR.

TallJimbo committed:
Fix discussion of compression.

- tiled table compression is supported by CFITSIO and other readers;

- we do need lossy compression and subtractive dithering for at least
  some image planes.

The overall result is that the big picture is not changed much: binary
tables are still not really viable, because they can't do the kind of
compression we need.

TallJimbo committed:
Include GPDF's lossy-compressed visualization-only image idea.

TallJimbo committed:
Add section on dual-interpretation HDUs.

TallJimbo committed:
Add note on third-party image support from GPDF.

TallJimbo committed:
Clarify situation on third-party viewers and per-HDU cells.

TallJimbo committed:
Clarify when overlap regions are and are not needed.

beckermr commented on index.rst:
```diff
@@ -44,19 +47,20 @@ Constraints from FITS
-- CFITSIO can do image tile compression, including efficient subimage reads that take advantage of the tile structure, but only on POSIX filesystems (or blocks of contiguous memory that correspond to the on-disk layout).
+- CFITSIO can do image tile compression and binary table compression, including efficient subimage reads that take advantage of the tile structure, but only on POSIX filesystems (or blocks of contiguous memory that correspond to the on-disk layout).
   It can also do subset reads of FITS binary tables, again only on POSIX filesystems.
+  CFITSIO can only permits customization of the quantization used in image tile compression via undocumented internal interfaces (which is what we use in ``afw``).
   This imposes the same limitation on its Python bindings in the ``fitsio`` package.
```
This is not totally true, though I agree the documentation is sparse.

beckermr replied:
We just added support for some of the lossy compression seeding options to fitsio: https://github.com/esheldon/fitsio/pull/411.

Excerpts from the docs are here in this github message: https://github.com/esheldon/fitsio/pull/411#issuecomment-2299028250

TallJimbo committed:
Mark technote as deprecated, since the problem is now easier.
